src/tool/abi/metadata_cache.cpp

```diff
@@ -276,163 +277,173 @@ void metadata_cache::process_class_dependencies(init_state& state, class_type& t
-        {
-            if (auto ifacePtr = checkRequiredInterface(ptr))
+            // Make sure that this interface reference applies for the same versioning "scheme" as the attribute
+            auto verMatch = match_versioning_scheme(attrVer, ifaceImpl);
+            if (!verMatch)
             {
-                exclusiveInterfaces.push_back(ifacePtr);
-            }
-        }
+                // No match on the interface reference is okay so long as there is _no_ versioning information on the
+                // reference. If there's not, then the requirement applies to all versioning schemes, so we look at the
+                // interface for the versioning information
+                if (get_attribute(ifaceImpl, metadata_namespace, "ContractVersionAttribute"sv) ||
+                    get_attribute(ifaceImpl, metadata_namespace, "VersionAttribute"sv))
+                {
+                    continue;
+                }
 
-        // Required interfaces of the exclusive interfaces need to be considered as well
-        for (std::size_t i = 0; i < exclusiveInterfaces.size(); ++i)
-        {
-            for (auto const& iface : exclusiveInterfaces[i]->type().InterfaceImpl())
-            {
-                auto dep = &find_dependent_type(state, iface.Interface());
-                if (auto ifacePtr = checkRequiredInterface(dep))
+                verMatch = match_versioning_scheme(attrVer, iface->type());
+                if (!verMatch)
                 {
-                    // It may be the case that we've already processed this type
-                    if (std::find(exclusiveInterfaces.begin(), exclusiveInterfaces.end(), ifacePtr) == exclusiveInterfaces.end())
-                    {
-                        exclusiveInterfaces.push_back(ifacePtr);
-                    }
+                    XLANG_ASSERT(false);
+                    continue;
                 }
             }
-        }
 
-        // We need to sort exclusive interfaces based off when they were added to the class. We accomplish this by
-        // looking at when the interface was introduced relative to the class' contract history (if it has one)
-        auto contractHistory = get_contracts(type.type());
-        auto contractPower = [&](interface_type const* iface)
-        {
-            auto initialContract = initial_contract(iface->type());
-            if (!initialContract)
+            // Take note if this is the default interface
+            if (is_default(ifaceImpl))
             {
-                xlang::throw_invalid("Interface '", iface->clr_full_name(), "' has no contract attribute, but is "
-                    "trying to be used as a fast ABI interface for class '", type.clr_full_name(), "' which does have "
-                    "a contract attribute");
+                XLANG_ASSERT(!fastInterface);
+                XLANG_ASSERT(!iface->fast_class);
+                fastInterface = const_cast<interface_type*>(iface);
```

Comment:
>iface [](start = 60, length = 5)

just declare iface non-const to begin with?

Reply:
It's not that easy since `find_dependent_type` returns a `const` ref since some of the items _are_ `const` and should be (e.g. element types, etc.). I may be able to do a bit better here since `interface_type`s are a little easier to look up

Reply:
On second thought, just going to add a comment. The `const_cast` here actually _is_ the behavior we want. We're getting something that _might_ be `const`, but when we reach this point, we _know_ that it's not.